source/bigint.c

```diff
@@ -448,254 +456,287 @@ int aws_bigint_bytebuf_debug_output(const struct aws_bigint *bigint, struct aws_
     }
+    buffer->len += byte_length;
 
-    uint8_t digit_byte = 0;
-    struct aws_byte_cursor digit_bytes_cursor = {
-        .ptr = &digit_byte,
-        .len = 1,
-    };
-
-    bool should_write = false;
-
-    for (size_t i = 0; i < digit_count; ++i) {
-        size_t digit_index = digit_count - i - 1;
-
-        uint32_t current_digit = 0;
-        aws_array_list_get_at(&bigint->digits, &current_digit, digit_index);
-
-        for (size_t j = 0; j < BYTES_PER_BASE_DIGIT; ++j) {
-            digit_byte = (uint8_t)(current_digit >> (3 * BITS_PER_BYTE)) & ((1U << BITS_PER_BYTE) - 1);
-            current_digit <<= BITS_PER_BYTE;
+    size_t bytes_to_write = byte_length;
+    if (bytes_to_write > digit_count * BYTES_PER_BASE_DIGIT) {
+        bytes_to_write = digit_count * BYTES_PER_BASE_DIGIT;
+    }
 
-            if (!should_write) {
-                should_write = digit_byte > 0 || j + 1 == BYTES_PER_BASE_DIGIT;
-            }
+    /* dumb/wasteful but simple */
+    size_t buffer_offset = byte_length - bytes_to_write;
+    uint32_t *raw_digits = bigint->digits.data;
+    for (size_t i = 0; i < bytes_to_write; ++i) {
+        size_t digit_index = i / BYTES_PER_BASE_DIGIT;
+        uint32_t digit = raw_digits[digit_index];
+        uint32_t digit_byte_index = i % BYTES_PER_BASE_DIGIT;
+        uint8_t digit_byte = (digit >> (BITS_PER_BYTE * digit_byte_index)) & 0xFF;
+        size_t buffer_index = bytes_to_write - i - 1 + buffer_offset;
 
-            if (should_write) {
-                if (aws_byte_buf_append(buffer, &digit_bytes_cursor)) {
-                    goto done;
-                }
-            }
-        }
+        raw_buffer[buffer_index] = digit_byte;
     }
 
     result = AWS_OP_SUCCESS;
 
 done:
 
-    AWS_POSTCONDITION(aws_bigint_is_valid(bigint));
-    AWS_POSTCONDITION(aws_byte_buf_is_valid(buffer));
+    AWS_FATAL_POSTCONDITION(s_aws_bigint_is_valid(bigint));
+    AWS_FATAL_POSTCONDITION(aws_byte_buf_is_valid(buffer));
 
     return result;
 }
 
-#if _MSC_VER
-#    pragma warning(pop)
-#endif
-
 bool aws_bigint_is_negative(const struct aws_bigint *bigint) {
-    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
 
     return bigint->sign < 0;
 }
 
+static int8_t s_aws_bigint_is_zero(const struct aws_bigint *bigint) {
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
+
+    volatile int8_t is_zero = 1;
+    size_t digit_count = aws_array_list_length(&bigint->digits);
+    uint32_t *raw_digits = bigint->digits.data;
+
+    for (size_t i = 0; i < digit_count; ++i) {
+        int64_t digit = (int64_t)raw_digits[i];
+        is_zero &= ((digit - 1) >> 63) & 0x01;
+    }
+
+    return is_zero;
+}
+
 bool aws_bigint_is_positive(const struct aws_bigint *bigint) {
-    AWS_PRECONDITION(aws_bigint_is_valid(bigint));
+    AWS_FATAL_PRECONDITION(s_aws_bigint_is_valid(bigint));
+
+    volatile uint8_t sign_one = 1 - ((bigint->sign >> 1) & 0x01);
```

Comment:
why is this volatile?

Reply:
wanted to prevent an early out that would skip the function call.

Reply:
reworked slightly to evaluate both conditions into volatile and then & the two at the end.